Yona/fabric_script.sh

```diff
@@ -0,0 +1,23 @@
+FABRIC_APIKEY_FILE="${SRCROOT}/fabric.apikey"
+FABRIC_BUILDSECRET_FILE="${SRCROOT}/fabric.buildsecret"
+
+if test ! -f "$FABRIC_APIKEY_FILE" -o ! -f "$FABRIC_BUILDSECRET_FILE"; then
+# Let the build fail
+exit 1
+fi
+
+FABRIC_APIKEY=$(cat "$FABRIC_APIKEY_FILE")
+if test $? -ne 0; then
+echo "Cannot read $FABRIC_APIKEY_FILE"
+exit 1
+fi
+
+FABRIC_BUILDSECRET=$(cat "$FABRIC_BUILDSECRET_FILE")
+if test $? -ne 0; then
+echo "Cannot read $FABRIC_BUILDSECRET_FILE"
+exit 1
+fi
+
+echo "Uploading dSYM files to Crashlytics"
+"${PODS_ROOT}/Fabric/run" "$FABRIC_APIKEY" "$FABRIC_BUILDSECRET"
```

Comment:
All of the above isn't necessary. In Jenkins, define two credentials of the kind "Secret text", one for the API key and one for the build secret (through [this URL](https://yonadev.ci.cloudbees.com/credentials/store/system/domain/_/newCredentials)). The use then in the Jenkinsfile as described [here](https://jenkins.io/doc/pipeline/steps/credentials-binding/). This will lead to something like this in the Jenkinsfile:
```
withCredentials([string(credentialsId: 'FabricApiKey', variable: 'FABRIC_API_KEY'), string(credentialsId: 'FabricBuildSecret', variable: 'FABRIC_BUILD_SECRET')]) {
  sh '''
  set +x
  "${PODS_ROOT}/Fabric/run" "$FABRIC_API_KEY" "$FABRIC_BUILD_SECRET"
  '''
}
```

On your developer machine, simply set these environment variables for your local builds, preferably with a different API key, to keep the secrets truly secret.

This way, the files with the secrets, the lines in ``.gitignore` and ``fabric_script.sh`` are not needed.

Reply:
We need to initialise the Crashlytics framework with API key on every Application launch. We should  either supply through the info.plist or from a local file in bundle like Fabric.apikey. Without this key Crashlytics will not know where to upload the Crash report.

Reply:
Sure. The original script was setting the environment variables by reading the files. Now with ``withCredentials``, Jenkins is setting the environment variables. That gives the same effect, in a better way.

Reply:
The API Secret and API Key is used by script to upload the .DSYM files to the portal only at the time of building the application. This will be one time process.

When the crash happens in device, the crash report is uploaded to the portal by the application with given API key in the bundle. For this, API Key is used inside Swift code to intialize the Crashlytics singleton during every launch of application by user. Then only it will know where to upload the crash report when crash happens.

The .DSYM file which is already uploaded to portal will be used process the crash report.

Reply:
I'm not getting the point here. You need to call ``${PODS_ROOT}/Fabric/run`` with the key and the secret, right? For these, you need environment variables and ``withCredentials`` is the appropriate way to set these variables.

Or you mean something else?

Reply:
func initializeCrashlytics(){
        var fabricAPIKey: String? = nil
        if let filepath = Bundle.main.path(forResource: "fabric.apikey", ofType: nil) {
            do {
                fabricAPIKey = try String(contentsOfFile: filepath)
            } catch {
                print(error.localizedDescription)
            }
        } else {
            print("File not found")
        }
        let whitespaceToTrim = CharacterSet.whitespacesAndNewlines
        let fabricAPIKeyTrimmed = fabricAPIKey?.trimmingCharacters(in: whitespaceToTrim)
        Crashlytics.start(withAPIKey: fabricAPIKeyTrimmed!)

This the code snippet used to initialize crashlytics. For this we need API key inside info.plist or in a file which will be bundle along with ipa files.

Reply:
Do you mean the procedure given by @aristotelos [here](https://github.com/yonadev/yona-app-ios/pull/386#discussion_r205672584) is not working at all?

Reply:
No Bert, that is working. We are using the same reference. Point 6 of procedure says below
```
Replace the Fabric startup code in your application delegate with the following code snippet that launches Crashlytics:
NSURL* resourceURL = [[NSBundle mainBundle] URLForResource:@"fabric.apikey" withExtension:nil];
NSStringEncoding usedEncoding;
NSString* fabricAPIKey = [NSString stringWithContentsOfURL:resourceURL usedEncoding:&usedEncoding error:NULL];

// The string that results from reading the bundle resource contains a trailing
// newline character, which we must remove now because Fabric/Crashlytics
// can't handle extraneous whitespace.
NSCharacterSet* whitespaceToTrim = [NSCharacterSet whitespaceAndNewlineCharacterSet];
NSString* fabricAPIKeyTrimmed = [fabricAPIKey stringByTrimmingCharactersInSet:whitespaceToTrim];

[Crashlytics startWithAPIKey:fabricAPIKeyTrimmed];

````

Reply:
So you are saying it worked with ``fabric_script.sh`` but doesn't work with ``withCredentials``?

If it worked with ``fabric_script.sh``, you should be able to make it work with ``withCredentials``, as that's just a different way to set the environment variable.

Reply:
``` fabric_script.sh ``` will work for the .DSYM file upload step. But API key is also used runtime by the application (after downloaded in the user device) to initialise the Crashlytics classes.

```
The Fabric API key: This is required at build time so that the build process can upload the debug symbols that later are required to symbolicate crash reports. The API key is also required at runtime so that the app on a customer device can submit crash reports.
The Fabric build secret: This is also required for uploading debug symbols at build time.
```

Reply:
That part doen't work when using the procedure given by @aristotelos [here](https://github.com/yonadev/yona-app-ios/pull/386#discussion_r205672584)?

Reply:
As per the procedure mentioned by @aristotelos here? we should use API key in both script and code.

-- Script will use the key & secret to upload DSYM files. Xcodebuild step will run this script from the file ```fabric_script.sh```  in the git repo. 

-- But Jenkins Environment variable cannot be used by the application which is running in the device. So we need to bundle the apikey file along with the other sources in application bundle (.ipa).

Reply:
I see what you mean. The API key is not only needed for the pod run command but it's also needed at runtime in the code. The original approach referenced by Aris was using the file for two purposes. Now I request to remove the file, so the other purpose isn't usable anymore either.

So you'll need to create the file build-time by echoing the environment variable ``FABRIC_API_KEY`` into the file. This requires you to rebuild a part of what I just asked you to remove. Sorry about that.

You'll need to add a reference to the file ``fabric.apikey`` again and add it to the ``.gitignore`` again. Then the build procedure should contain a construct like this:

```
withCredentials([string(credentialsId: 'FabricApiKey', variable: 'FABRIC_API_KEY'),] {
  sh '''
  set +x
  "echo "$FABRIC_API_KEY" >"fabric.apikey"
  '''
}
```

Reply:
Remove this file, as it is not necessary anymore.